### PR TITLE
Feature/base physics

### PR DESCRIPTION
**Pull request by Mefiso**
We can't include Leaks.h in ModulePhysics.cpp at the moment, because there are classes of Bullet Library that override the new operator and that generates conflicts with our MY_DEBUG macro. For more information about this issue ask me (Pol -Mefiso-) or Jordi.

In debug you'll notice some memory leaks detected by CRT library if you click the play button. Those leaks are generated in the stepSimulation() of Bullet library, not our code, and they do not appear when you execute Tesseract in Release. We are not even sure that they are actual leaks.

### CODE REVIEW

File: Game/Assets/Scenes/testLayers.scene.meta [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+    "Resources": [
+        {
+            "Id": 499249411013895419,
+            "Type": "Scene"
+        }
+    ],
+    "Timestamp": 1621516115
```
Comment (telmiuse) [inline Game/Assets/Scenes/testLayers.scene.meta:8]: Timestamp is not longer used.

File: Game/Assets/Scripts/ControllerTest.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+void ControllerTest::Start() {
+	kinematic = GameplaySystems::GetGameObject(kinematicUID);
+	if (kinematic) transform = kinematic->GetComponent<ComponentTransform>();
+}
+
+void ControllerTest::Update() {
+	if (Input::GetKeyCode(Input::KEYCODE::KEY_I)) {
+		transform->SetGlobalPosition(transform->GetGlobalPosition() + float3(0.f, 0.f, 1.f) * 4.f * Time::GetDeltaTime());
+		Debug::Log("GG!");
```
Comment (telmiuse) [inline Game/Assets/Scripts/ControllerTest.cpp:23]: :/ gg

File: Game/Assets/Scripts/testCollision.cpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+void testCollision::Start() {
+	
+}
+
+void testCollision::Update() {
+	
+}
+
+void OnCollision() {
+    Debug::Log("Collision Detected");
```
Comment (telmiuse) [inline Game/Assets/Scripts/testCollision.cpp:22]: i dont really undestand how this work couse i dont see some function in the gameplaysistem.cpp or includes

Comment (telmiuse) [reply Game/Assets/Scripts/testCollision.cpp:22]: is just to know for me .

File: Game/Assets/Scripts/ControllerTest.cpp (hunk 2 above)
Comment (Mefiso) [reply Game/Assets/Scripts/ControllerTest.cpp:23]: Both scripts and scene files won´t go to develop... They are here so the QA reviewer can perform the appropriate tests...

File: Game/Assets/Scripts/testCollision.cpp (hunk 3 above)
Comment (Mefiso) [reply Game/Assets/Scripts/testCollision.cpp:22]: So, the way collision reaction is handled by scripts is analogous to that of button clicked: the Script class has an onCollision method that may be overrided by your script; collider components also have a method onCollision, which gets all the ComponentScript in their owners (if any) and call their onCollision methods in turn.

Then, the flow is: ModulePhysics, on PreUpdate, performs the stepSimulation of the world; for every contact that occurred (if any), the onCollision method of the collider components involved is called, and they call the onCollision methods of the scripts owned by their owner.

Comment (telmiuse) [reply Game/Assets/Scripts/testCollision.cpp:22]: i got it, okay, just remove the timestamp and for me its allight

Comment (telmiuse) [reply Game/Assets/Scripts/testCollision.cpp:22]: Isaac should do some QA but this is up to Jaime..

**Review by telmiuse — DISMISSED** [review]

File: Game/Assets/Scripts/testCollision.cpp (hunk 3 above)
Comment (FrancPS) [reply Game/Assets/Scripts/testCollision.cpp:22]: Also, OnCollision is a virtual function in Script.h, that must be overriden in each script that need this functionality.

File: Project/Source/Modules/ModulePhysics.cpp [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+			/*
+
+
+
+			Sphere bullet;
+			bullet.color = Green;
+			bullet.radius = 1.0f;
+			bullet.SetPos(App->camera->Position.x, App->camera->Position.y, App->camera->Position.z);
+
+			App->physics->AddBody(bullet)->Push(-App->camera->Z.x * 20.0f, -App->camera->Z.y * 20.0f, -App->camera->Z.z * 20.0f);*/
```
Comment (JaimeBea) [inline Project/Source/Modules/ModulePhysics.cpp:172]: There is some commented code in this function.

File: Project/Source/Modules/ModulePhysics.cpp [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+	sphereCollider->rigidBody->setUserPointer(sphereCollider);
+	AddBodyToWorld(sphereCollider->rigidBody, sphereCollider->colliderType, sphereCollider->layer);
+}
+
+btRigidBody* ModulePhysics::AddSphereBody(MotionState* myMotionState, float radius, float mass) {
+	btCollisionShape* colShape = new btSphereShape(radius);
+
+	btVector3 localInertia(0, 0, 0);
+	if (mass != 0.f)
+		colShape->calculateLocalInertia(mass, localInertia);
```
Comment (JaimeBea) [inline Project/Source/Modules/ModulePhysics.cpp:206]: One-line ifs in separate lines should be avoided. There are a few instances in this file.

File: Project/Source/Modules/ModulePhysics.h [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+class ModulePhysics : public Module {
+public:
+	// ------- Core Functions ------ //
+	bool Init() override;
+	//bool Start();
+	UpdateStatus PreUpdate();
+	UpdateStatus Update();
+	//UpdateStatus PostUpdate();
+	bool CleanUp();
+	//void ReceiveEvent(TesseractEvent& e);
```
Comment (JaimeBea) [inline Project/Source/Modules/ModulePhysics.h:54]: Unused commented functions.

**Review by JaimeBea — APPROVED** [review]

File: Project/Source/Modules/ModulePhysics.cpp (hunk 4 above)
Comment (FrancPS) [reply Project/Source/Modules/ModulePhysics.cpp:172]: This code should not be included, it was just for testing. I'll remove it now.

File: Project/Source/Modules/ModulePhysics.cpp (hunk 5 above)
Comment (FrancPS) [reply Project/Source/Modules/ModulePhysics.cpp:206]: Ah the ctrl+K+D

**Review by JaimeBea — APPROVED** [review]

**Review by telmiuse — APPROVED** [review]

**Review by telmiuse — APPROVED** [review]

**Review by JaimeBea — APPROVED** [review]

**Pull request merged**